src/search-backend/search-backend-factory.ts

```diff
@@ -0,0 +1,40 @@
+import { FulltextSearchBackend } from './fulltext-search-backend';
+import { MetadataSearchBackend } from './metadata-search-backend';
+import { SearchType } from '../search-type';
+import type { SearchBackendInterface } from './search-backend-interface';
+import type { SearchBackendOptions } from './search-backend-options';
+import { Memoize } from 'typescript-memoize';
+
+/**
+ * A factory to obtain the correct search backend for a given search type.
+ */
+export class SearchBackendFactory {
+  private constructor() {
+    //
+  }
+
+  /**
+   * Retrieve a search backend that can handle the given type of search.
+   * @param type The type of search that the backend needs to handle.
+   * @param options Options to pass to the search backend.
+   */
+  @Memoize((type: SearchType, options: SearchBackendOptions = {}) => {
+    // We can memoize backends based on their params, to avoid constructing redundant backends
```

Comment:
This should only memoize the MetadataSearchBackend objects themselves, so that we don't construct multiple copies of identical backend objects. So two factory calls with the same search type (e.g., metadata) and same options should only be memoized once and should return the same object. Note this doesn't have any effect on the search queries themselves, only the backends.